Accept a Clock in DefaultDefaultClockProvider's Primary Constructor

austinarbor opened the pull request:
# Description
If the application has a bean of type `Clock` available in the context, pass that through to the `DefaultClockProvider` instead of using `Instant.now()`

This allows easier synchronization (and unit testing) when an application has a clock defined in the context

nakamura-to commented:
Thank you for your feedback!  
Let me clarify a bit.  

Komapper already has an interface called `ClockProvider` that allows you to use any `Clock` you want.  
You can also override the `ClockProvider` as a bean.  

Is there a specific reason why you prefer not to use `ClockProvider`?

austinarbor commented:
I am using `ClockProvider` now, but with `ClockProvider` you have to do something like
```kotlin
@Bean
internal fun clockProvider(clock: Clock): ClockProvider {
  return object : ClockProvider {
    override fun now(): Clock = Clock.fixed(clock.instant(), clock.zone)
  }
}
```

This works and isn't too bad, but IMO it's not that intuitive that you need to delegate your clock to a `Clock.fixed`.  I think checking for a Clock bean (if it exists) and using that will provide a nice layer of abstraction from the user.

If you think `ClockProvider` now is sufficient alone I am ok with that too

GreyTeardrop commented:
I wonder if it's expected for `ClockProvider` to necessary return a clock with a fixed time? 🤔 
Right now we return the same `clock` bean from all the calls to `ClockProvider.now()`
```kotlin
@Bean
fun komapperClockProvider(clockProvider: ObjectProvider<Clock>): ClockProvider {
    val clock = clockProvider.getIfAvailable() ?: Clock.systemUTC()
    return ClockProvider { clock }
}
```

austinarbor commented:
@nakamura-to i think that's a good compromise - but in that case I think we should also make `DefaultClockProvider` public so it can be easily used. I will make that change

austinarbor commented:
@nakamura-to change has been made and I added some tests... looks like the labeler needs to re-run or need to manually remove the two autoconfigure labels that were added from the original PR diff

nakamura-to commented:
@austinarbor  Thank you for your change. I have removed the two labels.
